mlir/include/mlir/Dialect/GPU/IR/GPUOps.td

```diff
@@ -1371,11 +1372,25 @@ def GPU_BarrierOp : GPU_Op<"barrier"> {
     accessing the same memory can be avoided by synchronizing work items
     in-between these accesses.
 
+    The address space of visible memory accesses can be modified by adding a
+    list of address spaces required to be visible. By default all address spaces
+    are included.
```

Comment:
What does it mean for memory to be visible? What memory accesses is this referencing?

Reply:
I might suggest "completed" or "committed" as synonyms here, if I've understood the semantics right.

That is, any memory operations involving the given memory spaces that were issued before the barrier must have their results reflected after the barrier completes.

Or, another way, any operation touching the listed memory spaces must be happens-before with any operation on those memory spaces after the barrier.

(This would allow us to not necessarily have `amdgpu.lds_barrier` - since we'd be able to express that as `gpu.barrier [#gpu.address_space<workgroup>]`

Reply:
I'm not trying to say that the memory is visible, but that the memory accesses are visible. I think it makes sense when paired with the previous paragraph
> and all memory accesses made by these work items prior to the op are
    visible to all work items in the workgroup.

I think what @krzysz00 is saying matches.

I'm open to suggestions on how to make it more clear.

Reply:
Some good points there, @krzysz00. I only know the amdgpu implementation and that's why I'd like to make sure the semantics make sense for everyone in a portable way.

The way I understand it, it's backend compiler's job to track any memory dependencies / synchronization within the same subgroup, and we only need `gpu.barrier` (the plain) version when the dependency can be across subgroups.

> (This would allow us to not necessarily have amdgpu.lds_barrier - since we'd be able to express that as gpu.barrier [#gpu.address_space<workgroup>]

I don't see how this aligns with the stated goal in the RFC:
> This could reduce the cost of synchronization.

> All work-items in the workgroup are still required to reach the barrier, but the address space visibility can be reduced.

lds_barrier is strictly more work than just gpu.barrier, no? It's `s_barrier` and 'please flush the shared memory fifo' . At least that's what c++ libraries do: https://github.com/ROCm/composable_kernel/blob/de3e3b642402eac5b4a466f6a2fa5e9f022ba680/include/ck/utility/synchronization.hpp#L20-L25.

Maybe I'm missing something.

Reply:
`gpu.barrier` is stronger than `amdgpu.lds_barrier`. The current semantics of `gpu.barrier` are not AMD's `s_barrier`, they're `(atomic fence) + s_barrier`.

(I think there are are even explicit calls to LLVM memory fencing in the AMD lowering path for `gpu.barrier`).

That is, `gpu.barrier` imposes a happens-before over all GPU memory by default, and this patch lets people loosen that.

Reply:
(and those libraries are implementing `amdgpu.lds_barrirer`. If they wanted to match `gpu.barrier`, they'd be waiting on `vmcnt(0)` as well

Reply:
> I'm not trying to say that the memory is visible, but that the memory accesses are visible.

I don't know what this means either. I don't want to be pedantic here, but this is just not the wording I'm used to / have seen in this context. I did some searching and found this thread that talks about something adjacent: memory access availability vs. visibility https://community.khronos.org/t/difference-between-availability-and-visibility/7401/2.

> I'm open to suggestions on how to make it more clear.

Maybe something like this (I reworded a few places):
```
    The "barrier" op synchronizes all work items of a workgroup. It is used
    to coordinate communication between the work items of the workgroup.
    `gpu.barrier` waits until all work items in the workgroup have reached this point
    and all memory accesses made by these work items prior to the op have
    completed and are visible to all work items in the workgroup.
    
    Data races between work items accessing the same memory locations can
    be avoided by synchronizing work items in-between these accesses.

    By default, all address spaces are included in `gpu.barrier`. This is equivalent to:
    ```mlir
    gpu.barrier memfence [#gpu.address_space<global>,
                          #gpu.address_space<workgroup>,
                          #gpu.address_space<private>]
    ```
    
    The `memfence` argument weakens the synchronization requrements such that
    only the listed address spaces need to have their accesses completed and visible across the
    workgroup. 
```

Reply:
I think that description basically covers it, but I would move a couple things around. Also I wouldn't say that `memfence` necessary weakens the memory fencing requirement, but it can be used for that.